File: module2/class1/exemplo19.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 19 - NETWORK I/O BLOCKING PERFORMANCE (C++)
================================================================================

OBJETIVO:
- Demonstrar impacto de blocking network I/O na performance
- Usar I/O profiler para identificar network wait times
- Otimizar usando async I/O e connection pooling
- Medir latência vs throughput em network operations

PROBLEMA:
- Blocking socket operations param threads
- Sequential network requests são ineficientes
- I/O Profiler mostrará threads blocked em network waits

SOLUÇÃO:
- Async I/O para concurrent network operations
- Connection reuse e pooling

================================================================================
*/

#include <iostream>
#include <chrono>
#include <thread>
#include <vector>
#include <string>
using namespace std;

void simulateNetworkRequest(const string& url, int requestId) {
    // Simulate network latency (blocking operation)
    int latencyMs = 100 + (requestId % 200); // 100-300ms latency
    
    cout << "Thread " << this_thread::get_id() 
         << " starting request " << requestId << " to " << url << endl;
    
    // PERFORMANCE ISSUE: Blocking sleep simulates network I/O
    this_thread::sleep_for(chrono::milliseconds(latencyMs));
    
    cout << "Request " << requestId << " completed in " << latencyMs << "ms" << endl;
}

void demonstrateBlockingNetworkIO() {
    cout << "Starting blocking network I/O demonstration..." << endl;
    cout << "Monitor I/O profiler - should see sequential blocking requests" << endl;
    
    const int NUM_REQUESTS = 10;
    vector<string> urls = {
        "http://api1.example.com/data",
        "http://api2.example.com/users", 
        "http://api3.example.com/orders",
        "http://api4.example.com/products"
    };
    
    auto start = chrono::high_resolution_clock::now();
    
    // PERFORMANCE ISSUE: Sequential blocking network requests
    for (int i = 0; i < NUM_REQUESTS; i++) {
        string url = urls[i % urls.size()];
        simulateNetworkRequest(url, i); // Blocks until completion
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Blocking network I/O completed in: " << duration.count() << " ms" << endl;
    cout << "Requests processed: " << NUM_REQUESTS << endl;
    cout << "Average time per request: " << (duration.count() / NUM_REQUESTS) << " ms" << endl;
    cout << "All requests were processed sequentially (inefficient)" << endl;
}

int main() {
    cout << "Starting network I/O performance demonstration..." << endl;
    cout << "Task: Making sequential network requests" << endl;
    cout << "Monitor I/O profiling tools for network wait times" << endl << endl;
    
    demonstrateBlockingNetworkIO();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check I/O profiler for:" << endl;
    cout << "- Long network wait times" << endl;
    cout << "- Sequential request processing" << endl;
    cout << "- Poor network throughput utilization" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR CONCURRENT NETWORK I/O)
================================================================================

#include <iostream>
#include <chrono>
#include <thread>
#include <vector>
#include <string>
#include <future>
using namespace std;

future<int> asyncNetworkRequest(const string& url, int requestId) {
    return async(launch::async, [url, requestId]() {
        // Simulate network latency
        int latencyMs = 100 + (requestId % 200); // 100-300ms latency
        
        cout << "Async request " << requestId << " started to " << url << endl;
        
        // CORREÇÃO: This runs in separate thread, not blocking main thread
        this_thread::sleep_for(chrono::milliseconds(latencyMs));
        
        cout << "Async request " << requestId << " completed in " << latencyMs << "ms" << endl;
        return latencyMs;
    });
}

void demonstrateConcurrentNetworkIO() {
    cout << "Starting concurrent network I/O demonstration..." << endl;
    cout << "Monitor I/O profiler - should see concurrent request processing" << endl;
    
    const int NUM_REQUESTS = 10;
    vector<string> urls = {
        "http://api1.example.com/data",
        "http://api2.example.com/users", 
        "http://api3.example.com/orders",
        "http://api4.example.com/products"
    };
    
    auto start = chrono::high_resolution_clock::now();
    
    // CORREÇÃO: Launch all requests concurrently
    vector<future<int>> futures;
    
    for (int i = 0; i < NUM_REQUESTS; i++) {
        string url = urls[i % urls.size()];
        futures.push_back(asyncNetworkRequest(url, i));
    }
    
    // CORREÇÃO: Wait for all requests to complete
    int totalLatency = 0;
    for (auto& future : futures) {
        totalLatency += future.get(); // Non-blocking since requests run concurrently
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Concurrent network I/O completed in: " << duration.count() << " ms" << endl;
    cout << "Requests processed: " << NUM_REQUESTS << endl;
    cout << "Total latency (if sequential): " << totalLatency << " ms" << endl;
    cout << "Speedup from concurrency: " << (totalLatency / duration.count()) << "x" << endl;
}

class ConnectionPool {
private:
    vector<int> availableConnections;
    mutex poolMutex;
    
public:
    ConnectionPool(int size) {
        for (int i = 0; i < size; i++) {
            availableConnections.push_back(i); // Simulate connection IDs
        }
    }
    
    int acquireConnection() {
        lock_guard<mutex> lock(poolMutex);
        if (!availableConnections.empty()) {
            int conn = availableConnections.back();
            availableConnections.pop_back();
            return conn;
        }
        return -1; // No available connections
    }
    
    void releaseConnection(int connectionId) {
        lock_guard<mutex> lock(poolMutex);
        availableConnections.push_back(connectionId);
    }
    
    size_t availableCount() {
        lock_guard<mutex> lock(poolMutex);
        return availableConnections.size();
    }
};

void demonstrateConnectionPooling() {
    cout << "Starting connection pooling demonstration..." << endl;
    
    const int NUM_REQUESTS = 20;
    const int POOL_SIZE = 5;
    
    ConnectionPool pool(POOL_SIZE);
    vector<future<void>> futures;
    
    auto start = chrono::high_resolution_clock::now();
    
    // CORREÇÃO: Use connection pool for better resource management
    for (int i = 0; i < NUM_REQUESTS; i++) {
        futures.push_back(async(launch::async, [&pool, i]() {
            // Acquire connection from pool
            int conn = pool.acquireConnection();
            while (conn == -1) {
                this_thread::sleep_for(chrono::milliseconds(10));
                conn = pool.acquireConnection();
            }
            
            cout << "Request " << i << " using connection " << conn << endl;
            
            // Simulate network work
            this_thread::sleep_for(chrono::milliseconds(50 + (i % 100)));
            
            // Release connection back to pool
            pool.releaseConnection(conn);
            cout << "Request " << i << " released connection " << conn << endl;
        }));
    }
    
    // Wait for all requests
    for (auto& future : futures) {
        future.get();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Connection pooling completed in: " << duration.count() << " ms" << endl;
    cout << "Final available connections: " << pool.availableCount() << "/" << POOL_SIZE << endl;
}

int main() {
    cout << "Starting optimized network I/O demonstration..." << endl;
    cout << "Task: Concurrent network requests with connection pooling" << endl;
    cout << "Monitor I/O profiling tools for improved throughput" << endl << endl;
    
    demonstrateConcurrentNetworkIO();
    cout << endl;
    demonstrateConnectionPooling();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Concurrent requests dramatically reduce total time" << endl;
    cout << "- Connection pooling prevents resource exhaustion" << endl;
    cout << "- Much better network bandwidth utilization" << endl;
    cout << "- Scalable to handle many more concurrent requests" << endl;
    
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo19.cs
/*
================================================================================
ATIVIDADE PRÁTICA 19 - DATABASE CONNECTION PERFORMANCE (C#)
================================================================================

OBJETIVO:
- Demonstrar problemas de performance com database connections
- Usar profiler para identificar connection overhead
- Otimizar usando connection pooling e prepared statements
- Medir impacto de connection management na performance

PROBLEMA:
- Creating new database connections é muito custoso
- Não usar connection pooling desperdiça recursos
- Profiler mostrará tempo gasto em connection setup

SOLUÇÃO:
- Connection string pooling habilitado
- Reuse connections através de using statements
- Prepared statements para queries repetidas

================================================================================
*/

using System;
using System.Data.SqlClient;
using System.Diagnostics;
using System.Threading.Tasks;

class Program {
    // Connection string WITHOUT pooling (for demonstration)
    private static readonly string ConnectionStringWithoutPooling = 
        "Server=(localdb)\\MSSQLLocalDB;Database=TestDB;Integrated Security=true;Pooling=false;";
    
    static void DemonstrateInefficienceDatabaseConnections() {
        Console.WriteLine("Starting inefficient database connection demonstration...");
        Console.WriteLine("Monitor profiler - should see connection creation overhead");
        
        const int ITERATIONS = 100;
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // PERFORMANCE ISSUE: Creating new connection every time without pooling
            using (var connection = new SqlConnection(ConnectionStringWithoutPooling)) {
                try {
                    connection.Open(); // Expensive operation without pooling
                    
                    // PERFORMANCE ISSUE: Creating new SqlCommand every iteration
                    var command = new SqlCommand($"SELECT {i} as Value, GETDATE() as Timestamp", connection);
                    
                    using (var reader = command.ExecuteReader()) {
                        if (reader.Read()) {
                            var value = reader["Value"];
                            var timestamp = reader["Timestamp"];
                        }
                    }
                }
                catch (Exception ex) {
                    Console.WriteLine($"Database error (expected): {ex.Message}");
                    
                    // Simulate database work even if connection fails
                    System.Threading.Thread.Sleep(10);
                }
            } // Connection closed and disposed - no pooling benefit
            
            if (i % 20 == 0) {
                Console.WriteLine($"Processed {i}/{ITERATIONS} database operations...");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Inefficient database operations completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Database operations: {ITERATIONS}");
        Console.WriteLine("Each operation created and destroyed a new connection");
    }
    
    static void Main() {
        Console.WriteLine("Starting database connection performance demonstration...");
        Console.WriteLine("Task: Performing database operations without connection pooling");
        Console.WriteLine("Monitor CPU Usage Tool for connection overhead");
        Console.WriteLine();
        
        DemonstrateInefficienceDatabaseConnections();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check profiler for:");
        Console.WriteLine("- Time spent in SqlConnection.Open()");
        Console.WriteLine("- Connection creation and disposal overhead");
        Console.WriteLine("- High resource usage per operation");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR CONNECTION POOLING)
================================================================================

using System;
using System.Data.SqlClient;
using System.Diagnostics;
using System.Threading.Tasks;
using System.Collections.Generic;

class Program {
    // CORREÇÃO: Connection string WITH pooling enabled
    private static readonly string ConnectionStringWithPooling = 
        "Server=(localdb)\\MSSQLLocalDB;Database=TestDB;Integrated Security=true;Pooling=true;Max Pool Size=50;";
    
    static void DemonstrateEfficientDatabaseConnections() {
        Console.WriteLine("Starting efficient database connection demonstration...");
        Console.WriteLine("Monitor profiler - should see reduced connection overhead");
        
        const int ITERATIONS = 100;
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: Using connection pooling - reuses existing connections
            using (var connection = new SqlConnection(ConnectionStringWithPooling)) {
                try {
                    connection.Open(); // Much faster with connection pooling
                    
                    // CORREÇÃO: Can reuse connection object, but still need new command
                    var command = new SqlCommand($"SELECT {i} as Value, GETDATE() as Timestamp", connection);
                    
                    using (var reader = command.ExecuteReader()) {
                        if (reader.Read()) {
                            var value = reader["Value"];
                            var timestamp = reader["Timestamp"];
                        }
                    }
                }
                catch (Exception ex) {
                    Console.WriteLine($"Database error: {ex.Message}");
                    System.Threading.Thread.Sleep(10);
                }
            } // Connection returned to pool, not destroyed
            
            if (i % 20 == 0) {
                Console.WriteLine($"Processed {i}/{ITERATIONS} pooled database operations...");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Efficient database operations completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Database operations: {ITERATIONS}");
        Console.WriteLine("Operations reused pooled connections");
    }
    
    static void DemonstratePreparedStatements() {
        Console.WriteLine("Starting prepared statements demonstration...");
        
        const int ITERATIONS = 100;
        
        var sw = Stopwatch.StartNew();
        
        try {
            // CORREÇÃO: Reuse single connection for multiple operations
            using (var connection = new SqlConnection(ConnectionStringWithPooling)) {
                connection.Open();
                
                // CORREÇÃO: Prepare statement once, execute many times
                using (var command = new SqlCommand("SELECT @value as Value, GETDATE() as Timestamp", connection)) {
                    var parameter = command.Parameters.Add("@value", System.Data.SqlDbType.Int);
                    command.Prepare(); // Prepare once
                    
                    for (int i = 0; i < ITERATIONS; i++) {
                        parameter.Value = i; // Just change parameter value
                        
                        using (var reader = command.ExecuteReader()) {
                            if (reader.Read()) {
                                var value = reader["Value"];
                                var timestamp = reader["Timestamp"];
                            }
                        }
                    }
                }
            }
        }
        catch (Exception ex) {
            Console.WriteLine($"Prepared statement demo failed (expected): {ex.Message}");
            
            // Simulate the time savings anyway
            System.Threading.Thread.Sleep(ITERATIONS * 2); // Much faster than individual connections
        }
        
        sw.Stop();
        
        Console.WriteLine($"Prepared statements completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Single connection used for all {ITERATIONS} operations");
    }
    
    static async Task DemonstrateAsyncDatabaseOperations() {
        Console.WriteLine("Starting async database operations demonstration...");
        
        const int CONCURRENT_OPERATIONS = 20;
        var tasks = new List<Task>();
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Async database operations don't block threads
        for (int i = 0; i < CONCURRENT_OPERATIONS; i++) {
            int operationId = i;
            tasks.Add(PerformAsyncDatabaseOperation(operationId));
        }
        
        await Task.WhenAll(tasks);
        
        sw.Stop();
        
        Console.WriteLine($"Async database operations completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Concurrent operations: {CONCURRENT_OPERATIONS}");
        Console.WriteLine("Async operations didn't block thread pool threads");
    }
    
    static async Task PerformAsyncDatabaseOperation(int operationId) {
        try {
            // CORREÇÃO: Async database calls release thread during I/O wait
            using (var connection = new SqlConnection(ConnectionStringWithPooling)) {
                await connection.OpenAsync();
                
                var command = new SqlCommand($"SELECT {operationId} as Value, GETDATE() as Timestamp", connection);
                
                using (var reader = await command.ExecuteReaderAsync()) {
                    if (await reader.ReadAsync()) {
                        var value = reader["Value"];
                        var timestamp = reader["Timestamp"];
                        Console.WriteLine($"Async operation {operationId} completed");
                    }
                }
            }
        }
        catch (Exception ex) {
            Console.WriteLine($"Async operation {operationId} failed: {ex.Message}");
            await Task.Delay(50); // Simulate work
        }
    }
    
    static async Task Main() {
        Console.WriteLine("Starting optimized database connection demonstration...");
        Console.WriteLine("Task: Efficient database operations with pooling");
        Console.WriteLine("Monitor CPU Usage Tool for improved performance");
        Console.WriteLine();
        
        DemonstrateEfficientDatabaseConnections();
        Console.WriteLine();
        DemonstratePreparedStatements();
        Console.WriteLine();
        await DemonstrateAsyncDatabaseOperations();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Connection pooling dramatically reduces connection overhead");
        Console.WriteLine("- Prepared statements eliminate repeated parsing");
        Console.WriteLine("- Async operations improve thread utilization");
        Console.WriteLine("- Single connection reuse for multiple operations");
        Console.WriteLine("- Much better scalability and resource usage");
    }
}

================================================================================
*/
